[FSDP] continue if param not exist in sharded load

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109117
* __->__ #109116

If I add a param and then wrap with FSDP + load state dict, when
strict=False don't hard error here.

Differential Revision: [D49170812](https://our.internmc.facebook.com/intern/diff/D49170812/)

rohan-varma committed:
[FSDP] continue if param not exist in sharded load

If I add a param and then wrap with FSDP + load state dict, when
strict=False don't hard error here.

Differential Revision: [D49170812](https://our.internmc.facebook.com/intern/diff/D49170812/)

[ghstack-poisoned]

fegin commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -604,7 +604,12 @@ def _sharded_pre_load_state_dict_hook(
+            param = state_dict.pop(fqn_from_global_root)
+        except KeyError:
+            logger.warning(f"Did not find param with FQN {fqn_from_global_root}, skipping it. The weight will not be filled if you expect it to be.")
+            continue  # TODO: only don't crash for strict=False
```
The TODO is not necessary. The error will be delayed but will be captured by nn.Module.load_state_dict.

fegin replied:
We can add a TODO here to adding UTs, https://github.com/pytorch/pytorch/issues/109134.

rohan-varma committed:
Update on "[FSDP] continue if param not exist in sharded load"


If I add a param and then wrap with FSDP + load state dict, when
strict=False don't hard error here.

Differential Revision: [D49170812](https://our.internmc.facebook.com/intern/diff/D49170812/)

[ghstack-poisoned]

fegin reviewed (approved):
LGTM, please fix the linter error, avoiding using f string with logging.

rohan-varma committed:
Update on "[FSDP] continue if param not exist in sharded load"


If I add a param and then wrap with FSDP + load state dict, when
strict=False don't hard error here.

Differential Revision: [D49170812](https://our.internmc.facebook.com/intern/diff/D49170812/)

[ghstack-poisoned]

b-chu commented:
Would you be able to include this in 2.1.0? This is a critical fix for crashes

b-chu commented:
@atalman Would you be able to add this to 2.1.1?

rohan-varma commented:
@b-chu I believe that 2.1.0 is already released, so unfortunately I'm a bit late for that. 

cc @atalman @osalpekar @seemethere  - can we make sure this gets into 2.1.1?

osalpekar commented:
@rohan-varma Sure, it's been added to the release-tracker: https://github.com/pytorch/pytorch/issues/110961
